From 6a5b3d09c71f785c9fbd163fda61346bdf04f06e Mon Sep 17 00:00:00 2001
From: Tomoharu Fukawa <tomoharu.fukawa.eb@renesas.com>
Date: Fri, 26 May 2017 15:47:08 +0900
Subject: [PATCH 395/909] iommu/ipmmu-vmsa: Add override support for the actual
 number of MMU contexts

commit 65a249b57fd970db7eb140625d0558960f43e1eb from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Add a IPMMU_VMSA_CTX_NUM to override the number of MMUs contexts that
you want to use.

The specified number of MMU contexts are used for the Renesas
IPMMU driver and the remaining contexts are unused.

SoCs, which is not R-Car Gen3, have no effect on IPMMU_VMSA_CTX_NUM.

Signed-off-by: Tomoharu Fukawa <tomoharu.fukawa.eb@renesas.com>
Signed-off-by: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
Signed-off-by: Hai Nguyen Pham <hai.pham.ud@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/iommu/Kconfig      | 11 +++++++++++
 drivers/iommu/ipmmu-vmsa.c |  2 +-
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/iommu/Kconfig b/drivers/iommu/Kconfig
index e53b8d58b62c..1c778fa01756 100644
--- a/drivers/iommu/Kconfig
+++ b/drivers/iommu/Kconfig
@@ -289,6 +289,17 @@ config IPMMU_VMSA
 
 	  If unsure, say N.
 
+config IPMMU_VMSA_CTX_NUM
+	int "Input MMU number of MMU's really used"
+	depends on IPMMU_VMSA
+	range 1 8
+	default 8
+	help
+	  Specify the number of MMU contexts that you want to use.
+
+	  The specified number of MMU contexts are used for the Renesas
+	  IPMMU driver and the remaining contexts are unused.
+
 config SPAPR_TCE_IOMMU
 	bool "sPAPR TCE IOMMU Support"
 	depends on PPC_POWERNV || PPC_PSERIES
diff --git a/drivers/iommu/ipmmu-vmsa.c b/drivers/iommu/ipmmu-vmsa.c
index 867e82cd0d7f..40d675525622 100644
--- a/drivers/iommu/ipmmu-vmsa.c
+++ b/drivers/iommu/ipmmu-vmsa.c
@@ -932,7 +932,7 @@ static const struct ipmmu_features ipmmu_features_default = {
 static const struct ipmmu_features ipmmu_features_rcar_gen3 = {
 	.use_ns_alias_offset = false,
 	.has_cache_leaf_nodes = true,
-	.number_of_contexts = 8,
+	.number_of_contexts = CONFIG_IPMMU_VMSA_CTX_NUM,
 	.setup_imbuscr = false,
 	.twobit_imttbcr_sl0 = true,
 	.reserved_context = true,
-- 
2.17.1

